src/main/java/appeng/core/sync/packets/PacketNEIBookmark.java

```diff
@@ -0,0 +1,85 @@
+import net.minecraft.nbt.CompressedStreamTools;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.common.util.ForgeDirection;
+
+import appeng.api.networking.energy.IEnergySource;
+import appeng.api.networking.security.BaseActionSource;
+import appeng.api.storage.IMEMonitor;
+import appeng.api.storage.data.IAEItemStack;
+import appeng.container.implementations.ContainerMEMonitorable;
+import appeng.core.sync.AppEngPacket;
+import appeng.core.sync.network.INetworkInfo;
+import appeng.util.InventoryAdaptor;
+import appeng.util.Platform;
+import appeng.util.item.AEItemStack;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+public class PacketNEIBookmark extends AppEngPacket {
+
+    private ItemStack bookmarkItem;
+
+    // automatic
+    public PacketNEIBookmark(final ByteBuf stream) throws IOException {
+        final ByteArrayInputStream bytes = new ByteArrayInputStream(stream.array());
+        bytes.skip(stream.readerIndex());
+        final NBTTagCompound comp = CompressedStreamTools.readCompressed(bytes);
+        if (comp != null) {
+            this.bookmarkItem = ItemStack.loadItemStackFromNBT(comp);
+        }
+    }
+
+    // api
+    public PacketNEIBookmark(final NBTTagCompound bookmarkItemComp) throws IOException {
+        final ByteBuf data = Unpooled.buffer();
+
+        final ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+        final DataOutputStream outputStream = new DataOutputStream(bytes);
+
+        data.writeInt(this.getPacketID());
+
+        CompressedStreamTools.writeCompressed(bookmarkItemComp, outputStream);
+        data.writeBytes(bytes.toByteArray());
+
+        this.configureWrite(data);
+    }
+
+    @Override
+    public void serverPacketData(final INetworkInfo manager, final AppEngPacket packet, final EntityPlayer player) {
+        if (isInventoryFull(player)) return;
+        final EntityPlayerMP pmp = (EntityPlayerMP) player;
+        final Container con = pmp.openContainer;
+
+        if (con instanceof ContainerMEMonitorable monitorable) {
+            final IMEMonitor<IAEItemStack> monitor = monitorable.getMonitor();
+            if (monitor != null) {
+                final IEnergySource energy = monitorable.getPowerSource();
+                final BaseActionSource actionSource = monitorable.getActionSource();
+
+                final AEItemStack request = AEItemStack.create(bookmarkItem);
+                final IAEItemStack out = Platform.poweredExtraction(energy, monitor, request, actionSource);
+                if (out != null) {
+                    final InventoryAdaptor adp = InventoryAdaptor.getAdaptor(player, ForgeDirection.UNKNOWN);
+                    ItemStack outItem = out.getItemStack();
+                    adp.addItems(outItem);
+                }
+            }
+        }
+    }
+
+    private boolean isInventoryFull(final EntityPlayer player) {
+        return !Arrays.asList(player.inventory.mainInventory).contains(null);
```

Comment:
This approach makes it so that if you have an inventory with partially-sized stacks, you can't pull the items even if they would stack with those partially-sized stacks. Is that intended?
For example if my inventory is full but has a stack of 32 redstone, but I wanted pull 8 redstone, it would not pull at all even though I could fit the 8 redstone by merging the stacks

Reply:
Yes you're right, it was a lame solution to the problem. I submitted a method that checks if the bookmarkItem fits or partially fits